core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java

```diff
@@ -366,6 +365,11 @@ private void updateTaskStatus(TaskStatus taskStatus)
             default:
         }
 
+        // Only allow stage state to transition to RUNNING, FLUSHING or FINISHED state
+        // when allTasks list is complete.
+        // If scheduling of tasks completes and all tasks are already finished then
+        // stage state will also be updated by schedulingComplete method.
+        State stageState = stateMachine.getState();
         if (stageState == SCHEDULED || stageState == RUNNING || stageState == FLUSHING) {
             if (taskState == TaskState.RUNNING) {
```

Comment:
Why we don't have `TaskState.RUNNING` in switch from line 342? Still, I will repeat another question — why we don't test stage state within task state switch? Could we finish the stage if we received flushing task? If we are testing state of stage, should we do this in synchronized block?

Reply:
> Why we don't have TaskState.RUNNING in switch from line 342?

Stage cannot transition to `RUNNING` if it's not in `SCHEDULED` state. `RUNNING` implies `SCHEDULED` (it's further state).

> why we don't test stage state within task state switch?

I'm not sure how this would look like

> Could we finish the stage if we received flushing task?

nope

> If we are testing state of stage, should we do this in synchronized block?

nope. `StageStateMachine` is lock-free by default. States only transition "forward".

Task state change might effect on stage stage change. It's only important to run `stage change loop` AFTER task state is change or stage becomes `SCHEDULED`